## Proyecto Pandas 
#### Jacobo Maté

In [1]:
import numpy as np # Importamos paquetes
import pandas as pd

In [2]:
path = '../data/attacks.csv' # Importamos base de datos
data = pd.read_csv(path, encoding = "ISO-8859-1")

In [3]:
pd.set_option('display.max_columns', None)  # ver todas las columnas

pd.set_option('display.max_rows', None)     # ver todas las filas

In [4]:
data.head() # Mostramos la cabeza de los datos

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2  original order Unnamed: 22 Unnamed: 23  
0    2018.06.25          6303.0         NaN         NaN  
1    2018.06.18          6302.0         NaN         NaN  
2    2018.06.09          6301.0         NaN         NaN  
3    2018.06.08          6300.0         NaN         NaN  
4    2018.06.04          6299.0         NaN         NaN

#### ¿Qué podemos sacar de este head?
 * Tenemos varias columnas con título con mayúsculas y espacios, voy a cambiarlas todas a lower y hacer replace de espacios. Esto es útil de cara a más tarde poder usar punto en vez de corchetes y comillado al querer referirme a una columna

In [5]:
data.columns = [c.lower().replace(' ', '') for c in data.columns] # Paso a minúsculas y quito los espacios para poder usar puntos
data.columns

Index(['casenumber', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal(y/n)', 'time',
       'species', 'investigatororsource', 'pdf', 'hrefformula', 'href',
       'casenumber.1', 'casenumber.2', 'originalorder', 'unnamed:22',
       'unnamed:23'],
      dtype='object')

In [6]:
data.info() # Procedemos a imprimir el tipo de datos con el que vamos a trabajar

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   casenumber            8702 non-null   object 
 1   date                  6302 non-null   object 
 2   year                  6300 non-null   float64
 3   type                  6298 non-null   object 
 4   country               6252 non-null   object 
 5   area                  5847 non-null   object 
 6   location              5762 non-null   object 
 7   activity              5758 non-null   object 
 8   name                  6092 non-null   object 
 9   sex                   5737 non-null   object 
 10  age                   3471 non-null   object 
 11  injury                6274 non-null   object 
 12  fatal(y/n)            5763 non-null   object 
 13  time                  2948 non-null   object 
 14  species               3464 non-null   object 
 15  investigatororsourc

In [7]:
data.shape

(25723, 24)

Tenemos un total de 25723 filas - 24 columnas

#### ¿Qué podemos sacar ya del tipo de datos con los que vamos a trabajar?
* Tenemos un total de 22 columnas de tipo objeto . serán cadenas de caracteres
* Tenemos dos columnas de tipo float - columnas 'year' y 'originalorder'
* Es raro que un año sea de tipo float -> No va a haber años float tipo 2023.5


In [8]:
# Queremos cambiar el tipo de dato de year a int.
# Para ello primero comprobamos si hay algun nan con any() - Sacamos la suma de nulos que hay en cada columna
data.isnull().sum()

casenumber              17021
date                    19421
year                    19423
type                    19425
country                 19471
area                    19876
location                19961
activity                19965
name                    19631
sex                     19986
age                     22252
injury                  19449
fatal(y/n)              19960
time                    22775
species                 22259
investigatororsource    19438
pdf                     19421
hrefformula             19422
href                    19421
casenumber.1            19421
casenumber.2            19421
originalorder           19414
unnamed:22              25722
unnamed:23              25721
dtype: int64

* En este output podemos ver la gran cantidad de valores nulos que hay. A simple vista se ve que mínimo hay 17000 nulos por columna. Coincide que en varias columnas hay exactamente 19421 nulos. 
* Partiendo de la hipótesis de que esos 19421 datos nulos van a concidir en fila con los otros, voy a decidir borrar dichas filas. 
* Para ello voy a tomar como número de nulos mas restrictivo el más pequeño 17021 - columna número de caso (casenumber) - sacaré los índices de dichas filas y borraré del database sus filas.

In [9]:
indices_malos = data[data.casenumber.isna() == True].index # Aquí usamos un filtro para almacenar en la variable indices_malos los indices de las filas de la columna casenumber que sean nulos
indices_malos

Index([ 5488,  8702,  8703,  8704,  8705,  8706,  8707,  8708,  8709,  8710,
       ...
       25712, 25713, 25714, 25715, 25716, 25717, 25718, 25719, 25720, 25721],
      dtype='int64', length=17021)

* Voy a revisar individualmente el caso del índice 5488 porque el resto parece que son seguidas e intuyo que el resto de valores tb sean nulos para ese rango 8702-25721

In [10]:
data[data.index == 5488]

casenumber                  date    year      type country     area  \
5488        NaN  Reported 06-Sep-1905  1905.0  Provoked     USA  Florida   

                          location activity         name sex  age  \
5488  Fort Pierce, St Lucie County  Fishing  David Curry   M  NaN   

                                                 injury fatal(y/n) time  \
5488  Lacerations to leg from hooked shark PROVOKED ...          N  NaN   

     species            investigatororsource                     pdf  \
5488     NaN  Muncie Evening Press, 8/6/1913  1905.09.06.R-Curry.pdf   

                                            hrefformula  \
5488  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href  casenumber.1  \
5488  http://sharkattackfile.net/spreadsheets/pdf_di...  1905.09.06.R   

      casenumber.2  originalorder unnamed:22 unnamed:23  
5488  1905.09.06.R          815.0        NaN        NaN

In [11]:
# Como si que hay datos para este ataque me lo voy a quedar. 
# Sí que voy a borrar los datos de las filas correspondientes al rango 8702-25721.

data.drop([i for i in range(8702,25721)],axis=0).shape

(8704, 24)

In [12]:
# Antes de aplicar el cambio veo que esté haciéndolo bien imprimiendo el shape. Hemos bajado las filas a 8704 así que bien -> Procedo con el inplace=True
data.drop([i for i in range(8702,25722)],axis=0,inplace=True)
data.shape

(8703, 24)

In [13]:
data.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

In [15]:
data.isnull().sum()

casenumber                 1
date                    2401
year                    2403
type                    2405
country                 2451
area                    2856
location                2941
activity                2945
name                    2611
sex                     2966
age                     5232
injury                  2429
fatal(y/n)              2940
time                    5755
species                 5239
investigatororsource    2418
pdf                     2401
hrefformula             2402
href                    2401
casenumber.1            2401
casenumber.2            2401
originalorder           2394
unnamed:22              8702
unnamed:23              8701
dtype: int64

Volviendo al punto donde nos había extrañado que el tipo de `year` fuera float; si queremos cambiar el tipo de esta columna a int tendremos que cambiar los nans a un valor -> sino al aplicar el astype() nos saltará un error.

Para ello voy a proceder como antes: saco los indices de las filas donde se encuentren valores nulos en la columna `year` -> Exploro tabla para esos índices y si nada me indica lo contrario cambio ese nan a algo con lo que pueda trabajar

In [16]:
yearsnulos = data[data.year.isna() == True].index
yearsnulos

Index([  187,  6079,  6302,  6303,  6304,  6305,  6306,  6307,  6308,  6309,
       ...
        8693,  8694,  8695,  8696,  8697,  8698,  8699,  8700,  8701, 25722],
      dtype='int64', length=2403)

In [34]:
data.loc[yearsnulos].head(10)

casenumber                  date  year        type    country  \
187   2017.01.08.R  Reported 08-Jan-2017   NaN     Invalid  AUSTRALIA   
6079  1836.08.19.R  Reported 19-Aug-1836   NaN  Unprovoked    ENGLAND   
6302             0                   NaN   NaN         NaN        NaN   
6303             0                   NaN   NaN         NaN        NaN   
6304             0                   NaN   NaN         NaN        NaN   
6305             0                   NaN   NaN         NaN        NaN   
6306             0                   NaN   NaN         NaN        NaN   
6307             0                   NaN   NaN         NaN        NaN   
6308             0                   NaN   NaN         NaN        NaN   
6309             0                   NaN   NaN         NaN        NaN   

            area    location      activity          name  sex  age  \
187   Queensland         NaN  Spearfishing  Kerry Daniel    M   35   
6079  Cumberland  Whitehaven      Swimming         a boy    M  NaN   
6302         NaN         NaN           NaN           NaN  NaN  NaN   
6303         NaN         NaN           NaN           NaN  NaN  NaN   
6304         NaN         NaN           NaN           NaN  NaN  NaN   
6305         NaN         NaN           NaN           NaN  NaN  NaN   
6306         NaN         NaN           NaN           NaN  NaN  NaN   
6307         NaN         NaN           NaN           NaN  NaN  NaN   
6308         NaN         NaN           NaN           NaN  NaN  NaN   
6309         NaN         NaN           NaN           NaN  NaN  NaN   

                                      injury fatal(y/n) time     species  \
187   No attack, shark made a threat display        NaN  NaN  Bull shark   
6079                                   FATAL          Y  NaN         NaN   
6302                                     NaN        NaN  NaN         NaN   
6303                                     NaN        NaN  NaN         NaN   
6304                                     NaN        NaN  NaN         NaN   
6305                                     NaN        NaN  NaN         NaN   
6306                                     NaN        NaN  NaN         NaN   
6307                                     NaN        NaN  NaN         NaN   
6308                                     NaN        NaN  NaN         NaN   
6309                                     NaN        NaN  NaN         NaN   

        investigatororsource                           pdf  \
187   Liquid Vision 1/8/2017  2017.01.08.R-KerryDaniel.pdf   
6079          C. Moore, GSAF   1835.08.19.R-Whitehaven.pdf   
6302                     NaN                           NaN   
6303                     NaN                           NaN   
6304                     NaN                           NaN   
6305                     NaN                           NaN   
6306                     NaN                           NaN   
6307                     NaN                           NaN   
6308                     NaN                           NaN   
6309                     NaN                           NaN   

                                            hrefformula  \
187   http://sharkattackfile.net/spreadsheets/pdf_di...   
6079  http://sharkattackfile.net/spreadsheets/pdf_di...   
6302                                                NaN   
6303                                                NaN   
6304                                                NaN   
6305                                                NaN   
6306                                                NaN   
6307                                                NaN   
6308                                                NaN   
6309                                                NaN   

                                                   href  casenumber.1  \
187   http://sharkattackfile.net/spreadsheets/pdf_di...  2017.01.08.R   
6079  http://sharkattackfile.net/spreadsheets/pdf_di...  1836.08.19.R   
6302                                        

In [33]:
data.loc[yearsnulos].tail(10)

casenumber date  year type country area location activity name  sex  \
8693           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8694           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8695           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8696           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8697           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8698           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8699           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8700           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8701           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722         xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       age injury fatal(y/n) time species investigatororsource  pdf  \
8693   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8694   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8695   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8696   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8697   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8698   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8699   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8700   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8701   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
25722  NaN    NaN        NaN  NaN     NaN                  NaN  NaN   

      hrefformula href casenumber.1 casenumber.2  originalorder unnamed:22  \
8693          NaN  NaN          NaN          NaN            NaN        NaN   
8694          NaN  NaN          NaN          NaN            NaN        NaN   
8695          NaN  NaN          NaN          NaN            NaN        NaN   
8696          NaN  NaN          NaN          NaN            NaN        NaN   
8697          NaN  NaN          NaN          NaN            NaN        NaN   
8698          NaN  NaN          NaN          NaN            NaN        NaN   
8699          NaN  NaN          NaN          NaN            NaN        NaN   
8700          NaN  NaN          NaN          NaN            NaN        NaN   
8701          NaN  NaN          NaN          NaN            NaN        NaN   
25722         NaN  NaN          NaN          NaN            NaN        NaN   

      unnamed:23  
8693         NaN  
8694         NaN  
8695         NaN  
8696         NaN  
8697         NaN  
8698         NaN  
8699         NaN  
8700         NaN  
8701         NaN  
25722        NaN

* Vista la salida, intuyo que desde la fila 6303 hasta 8701 tenemos valores nulos en todas las columnas salvo en `originalorder`.
* Como de 6302 a 8701 hay 2400 filas y sabemos que en total hay 2403 nulos, podemos confirmar que ese rango lo podemos eliminar todo también.
* Si todos los valores son nulos menos una columna tampoco me sirven de mucho -> Puedo hacer un último checkeo para comprobar que mi hipótesis sea verdadera y que no me esté fumando datos correctos por ahí enmedio.

In [35]:
data.drop([i for i in range(6303,8701)],axis=0).shape

(6305, 24)

In [37]:
data.drop([i for i in range(6303,8701)],axis=0, inplace=True)
data.shape

(6305, 24)

In [38]:
data.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

In [40]:
data.isnull().sum()

casenumber                 1
date                       3
year                       5
type                       7
country                   53
area                     458
location                 543
activity                 547
name                     213
sex                      568
age                     2834
injury                    31
fatal(y/n)               542
time                    3357
species                 2841
investigatororsource      20
pdf                        3
hrefformula                4
href                       3
casenumber.1               3
casenumber.2               3
originalorder              2
unnamed:22              6304
unnamed:23              6303
dtype: int64

Después de borrar los nans del paso anterior vuelvo a mostrar la suma de estos.